### ISSUE TITLE
Enhance pyinstaller by splitting the hooks into separate repositories/packages

### ISSUE THREAD

**Issue by virtuald** [pyinstaller/pyinstaller#2692]
Problem
======

pyinstaller has (as of this writing) 41 open PRs of varying quality and 424 github issues, many of which seem to be hook related. As pyinstaller has grown in popularity, I'm sure the maintenance burden has grown and consumes a large amount of time for the maintainers. Additionally, I suspect that the current maintainers of pyinstaller are not experts most of the various packages that pyinstaller has hooks for, nor do they want to be.

Because of the maintenance burden, PRs take months to get merged, which is a huge problem in my opinion (and I'm sure others are tired of this as well). It is significantly hindering pyinstaller development (and probably adoption as well) -- and really, the pyinstaller maintainers don't have a good way to test said hooks nor the domain knowledge to actually tell whether a given hook is good or bad.

Proposed Solution: divide and conquer
============================

We split up pyinstaller development into many different packages that have their own git repo, and their own pypi packages. The core pyinstaller package will use [setuptools entry points](http://setuptools.readthedocs.io/en/latest/pkg_resources.html#entry-points) to find all of the various hooks. One advantage of this is that outside organizations could maintain and distribute their own set of hooks without needing to interact directly with pyinstaller developers (though, I think we should try to encourage the community to create hooks in official pyinstaller repos if possible).

This would ease the burden of maintaining pyinstaller on the current developers and allow greater community participation in the development of pyinstaller, by allowing people to focus on their particular interests -- for example, I know a ton about how PyGObject works and the things pyinstaller needs to do to hook it, but not really a lot about how the core of pyinstaller works; nor do I really care as long as it works. 

A notional refactoring would probably generate the following pypi packages:

* `pyinstaller-core`: all of the core logic that pyinstaller needs that isn't hook related
* `pyinstaller`: a virtual package that depends on pyinstaller-core and all of the 'official' pyinstaller hook packages
  * Maybe these dependencies are versioned, maybe not
* `pyinstaller-hooks-qt`
* `pyinstaller-hooks-pygobject`
* ... etc

Various community members would be given commit privileges to the hooks repos, and they would decide on when new releases are pushed to pypi *separately* from pyinstaller-core. Ideally, a new pypi release would be pushed via travis-ci's deploy mechanism whenever a new git tag is created.

Downsides
-----------

I'm sure there are more, but here's what I have so far:

* pyinstaller-core changes could break hooks without knowing it
   * Need to figure out a good way to test this
* Versioning of releases could be an issue
  * I imagine that `pyinstaller-core` would be released much less often than the various hooks repos

My opinion is that the benefits of splitting hooks from core significantly outweigh the downsides.

Tasks
------

* [ ] Implement the setuptools entry point mechanism to load hooks
* [ ] Hooks projects should all have the same structure, so we should implement an example hooks repository to be used as a template
  * Projects include tests, travis-ci/appveyor configurations
* [ ] Decide on the names/number of hooks repos
* [ ] Register the packages on pypi
* [ ] Extract the hooks and their relevant tests to the hooks repos
  * Ensure that hooks are only using documented APIs, or document the APIs they're using
  * Update hooks documentation (perhaps create a hook developer guide?)
* [ ] Move any hooks related issues to the hooks repos
* [ ] Setup travis-ci and appveyor tests for each hooks repo
* [ ] Setup auto-deploy mechanism for git tags

I estimate it would take a few months to get all of this done, but would result in a streamlined pyinstaller development process in the future.

Final Thoughts
===========

I will not personally have time until mid-august to contribute to this effort. However, if we do this, I could help with some of the tasks above, and I would volunteer to maintain the pyinstaller-hooks-pygobject project (including the hooks for Gtk, GStreamer, et al).

**Comment by xoviat** [comment]
The amount of time for pull requests to be merged has been unacceptable, but there is no alternative to PyInstaller.

**Comment by virtuald** [comment]
@xoviat this proposal doesn't get rid of pyinstaller, but decentralizes it so that we can make more progress as a community, instead of excessively burdening maintainers

**Comment by xoviat** [comment]
I am aware of that. My implication was that the only solution to this problem is to attempt to improve PyInstaller rather than starting another project.

**Comment by tallforasmurf** [comment]
> pyinstaller-core changes could break hooks without knowing it 

The hook API (pyinstaller.utils.hooks) needs to be much better documented, and all the current hooks inspected to ensure that all of their dependencies are included in that module. Any sneaky refs to things outside that module need to be found and eliminated, or their dependencies brought into the official API.

Probably a hook-developer manual needs to be written, replacing the current manual section (http://pyinstaller.readthedocs.io/en/stable/hooks.html).

**Comment by virtuald** [comment]
@tallforasmurf added your comments to the task list

**Comment by kelson42** [comment]
On a higher level, this seems to make sense. This is pretty sad to see a useful project where PRs are not merged quickly (and the product not released often).

A few words about the "Downsides". To avoid problems with hooks, this should be tested properly in the CI (this seems mandatory here). Regarding the versioning, I would recommend (at least at the beginning) to start with one release for all the modules = move/release all of them together.

**Comment by htgoebel** [comment]
TL;DR: This is the direction to go. Maybe some details need to be adjusted. Who will become the lead?

I agree to your analysis. This proposal will not solve all of the issue, esp. the poor quality of pull-requests and the lack of people contributing. But this proposal ought to solve the long release-cycles for hooks.

When I first read the proposal, I though: "Hey, why can't this be done within the main repo? Giving commit-permissions to hooks should not be much of a problem". After re-thinking, I found that this will still not solve the issue of long-release cycles. And reasons for the long release-cycles are (among others) a) stability of the code in the `develop` branch, b) the effort for releasing itself, and c) the work for writing the Changelog.

When decoupling the hooks from the core, we could at least reduce the burden of writing the change log a lot. And is has the additional benefit that the core-developers (well, currently it's only me left) can concentrate on the core features.

**So let's go for it.**

Some thoughts about the details:
* IMO, a single package should be enough.The lead-maintainers of this package can delegate hooks for specific packages. An organizational rule should be okay for this.
* Regarding entry-points:
  * If we have only one package, there is no need for implementing the entry-points mechanism – thus the implementation could be much simpler, we could get this go much quicker. (If required, entry-points can be added later.)
  * OTH using entry-points could allow to eliminate some code in PyInstaller (ModuleHookCache, HookCache). OTH again, maybe its not worth the effort.
* From pyinstaller's point of view hooks are *files*, not *modules*. Hooks are not imported into sys.modules, but just executed. (Even if the import machinery is used for this.) Thus hooks are resources which PyInstaller would list by simply using `pkg_resources.resource_listdir('pyinstaller-hooks', 'hooks')`
* Regarding the tests and requirements:
  * PyInstaller would simply require the `pyinstaller-hooks` package (without version). So at travis/appveyor this would always pull the latest release.
  * Most (if not all) "Library" tests would be removed from PyInstaller.
  * For pyinstaller-hooks CI we could have two cases: One using the current release and one using the latest development head.
  * pyinstaller-hooks would *not* recompile the bootloader. Changes to the bootloader rarely effect the collected modules, thus we can save this step.
  * Implementing this should be not much of a problem since most of this is already there.
* Transition: This ought to be developed on a branch by it's own (e.g. `external-hooks`). Until this is merged into `develop`, the `pyinstaller-hooks` CI would pull this branch.
* Others:
  * PyInstaller needs to log which version of `pyinstaller-hooks` was active. (When using entry-points: all of these packages plus version).

**License:**: I propose to use the same license as for PyInstaller itself (GPL plus bootloader exception). This avoids confusion be the users.

**Comment by htgoebel** [comment]
Addendum:
* Some hooks may still need to be delivered with pyinstaller, e.g. those related to pkg_resources.
* Due to this we *may* need to use the entry points in any case. 
* I suggest versions to be simply the release date.
* I suggest branch-management to be the same as for PyInstaller (developing on `develop`, releasing on `master`).

I created a repository for this: https://github.com/pyinstaller/pyinstaller-hooks, pushed the initial data and registered at [Test-PyPI](https://test.pypi.org/project/PyInstaller-Hooks/) and (mind the new URL) and [PyPI](https://pypi.python.org/pypi/PyInstaller-Hooks/).

**Comment by virtuald** [comment]
Thoughts:

* If you copy the existing hooks to another repo, they have to be GPL unless you get permission from all contributors, so that settles the license question :)

I feel fairly strongly that the hooks should be implemented via a dynamic discovery mechanism such as entry points, and that it should be split into N projects:

  * The entry points mechanism is very simple to setup/use
  * I envision that the entry point would actually be an object that pyinstaller would interact with to enumerate the hooks that the entry point provides, so it wouldn't matter whether they were modules or files or whatever
    * This could allow more interesting extensions of pyinstaller in the future
  * As a user that only cares about a *very* small subset of the available hooks, I don't want to see github issue noise from other hooks (for example -- I don't care about QT problems). I suspect this is true for most users, nobody uses all of the available hooks in a single application, and most hooks would fall into a particular category
  * There are some communities which may desire to create their own set of hooks independently from pyinstaller, and this is a convenient way to allow that to happen
  * I don't see why the pkg_resources hooks would have to be included in the core pyinstaller project

As I said before, I can dedicate time to this at the middle/end of August, but no sooner. I don't expect making it work would be all that much work; I think setting up repos/ci/tests and migrating issues would take up a lot more time

**Comment by htgoebel** [comment]
> I feel fairly strongly that the hooks should be implemented via a dynamic discovery mechanism such as entry points, 

Using entry-points is okay for me. I just wanted to avoid the implementation effort in the first step.

> and that it should be split into N projects:

From the user's perspective, in case of an problem, this may become troublesome. Where should the user report a bug or ask for help? This will be uncomfortable with just `pyinstaller-hooks`, but will be come complicated when splitting into even more packages.

> As a user that only cares about a very small subset of the available hooks, I don't want to see github issue noise from other hooks (for example -- I don't care about QT problems).

You can "mute" the ticket by a simple click. There are about 200 hook-related issues per year (rough estimate), which means about one each working-day. YMMV, but i think, this is manageable.

OTOH I would expect the hook-developers to monitor the main repo's issues anyway. If the hook developers do not monitor the issues here, we will generate a lot of "please report this at …" noise. And who is going to write all these replay? I'm afraid I'm the one carrying this burden.

If the noise for becomes to much, we could implement support for the mention-bot. Python did this: https://github.com/python/core-workflow/issues/32

**Comment by xoviat** [comment]
> When I first read the proposal, I though: "Hey, why can't this be done within the main repo? Giving commit-permissions to hooks should not be much of a problem". After re-thinking, I found that this will still not solve the issue of long-release cycles. And reasons for the long release-cycles are (among others) a) stability of the code in the develop branch, b) the effort for releasing itself, and c) the work for writing the Changelog.

a) is probably unsolvable but b) and c) are not. It should be possible to implement fully automated releases (just git tag && git push tags) combined with towncrier, which reduces the release effort to a few minutes at most.

**Comment by htgoebel** [comment]
Our release process is documented in https://github.com/pyinstaller/pyinstaller/wiki/Development-Release. It takes about 1/2 hour (beside writing the changelog), since it includes quiet some checks to avoid problems. If you have ideas how to simplify the process by keeping the quality, I'm open for suggestions.

Regarding towncrier: This too"only" creates a NEWS.rst file from some fragments which need to be created first. This does not help much, since the time-consuming work is to scan all changes, issues and pull-requests for relevant information.

**Comment by xoviat** [comment]
> Regarding towncrier: This too"only" creates a NEWS.rst file from some fragments which need to be created first. This does not help much, since the time-consuming work is to scan all changes, issues and pull-requests for relevant information.

Yes, well the fragments are created with each pull request rather than all at once. That moves effort onto contributors. You of course need a bot to check this (or we can just use a unit test for now).

**Comment by xoviat** [comment]
So let me write a checklist here (ordering to save the most effort first):

- [ ] Pin libraries so that the test suite works.
- [ ] Implement towncrier to build NEWS over time with each PR (implement unittest that fails if TRAVIS_PULL_REQUEST is present and news file is not present).
- [ ] Implement cron to recompile bootloaders and commit changes back to the repository automatically.
- [ ] Implement cron to automatically check tags and post new releases on web site.

**Comment by xoviat** [comment]
The bot is at pypa/browntruck

**Comment by htgoebel** [comment]
I think it's time to move this discussion to a ticket of it's own, since this very ticket is about caving out the hooks, no about the release-process.

Nevertheless I need to answer here :-)

> That moves effort onto contributors

I'm happy if I get good commit-messages, and IMHO if I need to plague the contributors, I'd focus on this. Otherwise I'd need to plague them with another text to be written well.

> Implement cron to recompile bootloaders and commit changes back to the repository automatically

I'm not trusting bootloaders build on some system we have no control over. No, I'm *not* trusting cloud services. Our users need to trust us so we need to take care that the bootloaders are not contaminated.

> Implement cron to automatically check tags and post new releases on web site.

Do you mean a cron to but the information about the new release on the web-site? This would save some minutes. If you mean a cron *creating* the release, the same is true as for bootloaders: Releases will not be build on some build-system in the cloud. Also the releases archives are signed and I'm not going to but any signing key on any server.

> The bot is at pypa/browntruck

A bot for which service? The repo has no description, no Readme, no website, nor any topics are provided. :-(

**Comment by htgoebel** [comment]
On the [debops/debop](http://github.com/debops/debops) mailinglist I just got this information. Maybe this is useful for us:

> https://help.github.com/articles/about-codeowners/.
In short, it's a file in the root directory of the repository that specifies
what files or paths are maintained by who, and when a new PR is published that
changes one of these files, a person is added as a reviewer of that PR. Of
course this works only on GitHub, but could be scripted elsewhere. Linux
kernel uses a similar system with the MAINTAINERS file, which was an
inspiration for the CODEOWNERS on GitHub.

And a reply on this:
> > That would solve it. [Being much more difficult to watch only parts of DebOps]
>
> Hm, at least partly. People will still open issues I guess which would notify
all people watching the repo. At least for that case, a issue template might be
useful.

**Comment by htgoebel** [comment]
I should have mentioned:
[Debops](https://debops.org) is a huge set of collaborating roles (aka recipes) for Ansible. debops currently discusses merging *all* "roles" into the main repository. The discussion can be found at https://lists.debops.org/pipermail/debops-users/2017-August/000078.html.

The roles have much more interdependences then out hooks and a role usually consists of about 10 files (incl. "recipe, templates, documentation) But the discussion about merging it into the main repo may give some insight.

**Comment by htgoebel** [comment]
After a deep discussion. nobody went ahead working on this. THus I assume there is need need (or at least not pressing need) ans will close this.
